### PR TITLE
Update Intake Review Endpoint to Support Different Claimant Types

### PR DESCRIPTION
**Pull request by jcq**
Connects #14505

### Description
This updates the `IntakesController#review` endpoint to support different claimant types. It also updates the controller schema for said endpoint.

### Acceptance Criteria
- [ ] Code compiles correctly

### CODE REVIEW

File: app/schemas/intakes_schemas.rb [hunk 1]
```diff
@@ -22,6 +22,8 @@ def review
                included_in?: DOCKET_OPTIONS,
                doc: "Appeals only"
         string :claimant, optional: true, nullable: true
+        string :claimant_type, optional: true, nullable: true
```
Comment (nanotone) [inline app/schemas/intakes_schemas.rb:25]: What do you think about dropping the `veteran_is_not_claimant` boolean in this schema, and deriving that from `claimant_type`? It might make some of the backend code cleaner, to not have to pass around an extra field, and not have to worry about inconsistencies.

I would also suggest slapping an `included_in?` restriction on this field, since it's essentially going to be an enum of four values. And if the RAMP code omits it (i.e. optional: true) then it could probably even be nullable: false.

(Insert some proverb about how the controller schema architect zooms in on the controller schema part of this PR 😛)

Comment (jcq) [reply app/schemas/intakes_schemas.rb:25]: Definitely a good idea re the `included_in?` restriction!

And I think the removal of `veteran_is_not_claimant` likely makes sense as well, and will have to see how extensive it is used

File: app/schemas/controller_schema.rb [hunk 2]
```diff
@@ -65,7 +65,7 @@ def initialize(format, &block)
 
   # mutates params by removing fields not declared in the schema, other than path params
   def remove_unknown_keys(params, path_params = {})
-    allowed = fields.map(&:name) + path_params.keys.map(&:to_s)
+    allowed = (fields.map(&:name) + path_params.keys).map(&:to_s)
```
Comment (nanotone) [inline app/schemas/controller_schema.rb:68]: Fixing a bug from #14628 where too many parameters were getting removed due to symbol/string mismatch.

File: app/schemas/intakes_schemas.rb [hunk 3]
```diff
@@ -15,18 +15,15 @@ def create
     def review
       ControllerSchema.json do
```
Comment (nanotone) [inline app/schemas/intakes_schemas.rb:16]: Couple notable changes in the controller schema for this endpoint:
- Because the frontend is doing a lot of the basic validation, and intelligently displaying the correct error messages, a lot of these fields that were nullable are now properly non-nullable. (previously, the frontend was relying on _model_ layer validations to get error messages, which was way too much coupling)
- `claimant_type` completely supercedes `veteran_is_not_claimant`. Thus, it should be impossible for a review's veteran_is_not_claimant to conflict with its claimant model's class type. This change, while logical here, has far-reaching consequences on the size of this PR 😅

File: app/models/decision_review_intake.rb [hunk 4]
```diff
@@ -52,6 +52,27 @@ def build_issues(request_issues_data)
 
   private
 
+  def create_claimant!
```
Comment (nanotone) [inline app/models/decision_review_intake.rb:55]: The `create_claimant!` methods in AppealIntake and ClaimReviewIntake were so similar, that I decided to refactor them up into the superclass.

This method preserves some of the trickiness that its predecessors relied on:
- If the user clicks submit review, then hits the back button, they can actually change the claimant type (e.g. from veteran to dependent), and successfully re-submit. Yeah, there's a feature test for this too.
- The claimant validator does a bunch of error-check (e.g. for address validity) but in order to report back to the frontend, it needs to set error states on the DecisionReview object, which then gets picked up by the Intake object when request finishes. I find this questionable, but we'll continue to live with it for now.

File: app/models/decision_review_intake.rb [hunk 5]
```diff
@@ -104,4 +125,34 @@ def after_validated_pre_start!
...
+  def participant_id
+    if %w[VeteranClaimant OtherClaimant].include? claimant_type
+      veteran.participant_id
+    else
+      request_params[:claimant]
+    end
+  end
+
+  # :nocov:
+  def review_param_keys
```
Comment (nanotone) [inline app/models/decision_review_intake.rb:148]: Another similarity between the Intake subclasses was how they unpacked (and permitted) parameters from the Rails request. This meant the model code was tightly coupled to HTTP request specifics, so I did a tiny bit of refactoring. Now the subclasses only specify which fields they need.

File: client/app/intake/actions/decisionReview.js [hunk 6]
```diff
@@ -12,6 +12,16 @@ export const submitReview = (intakeId, intakeData, intakeType) => (dispatch) =>
...
   });
 
+  const validationErrors = validateReviewData(intakeData, intakeType);
+  if (validationErrors) {
+    dispatch({
+      type: ACTIONS.SUBMIT_REVIEW_FAIL,
+      payload: { responseErrorCodes: validationErrors },
+      meta: { analytics }
+    });
+    return Promise.reject();
```
Comment (nanotone) [inline client/app/intake/actions/decisionReview.js:22]: This is where frontend validation happens, and if anything is obviously missing, skip the XHR and go straight to showing error messages.

File: client/app/intake/components/SelectClaimant.jsx [hunk 7]
```diff
@@ -82,19 +83,26 @@ export const SelectClaimant = (props) => {
     veteranIsNotClaimant,
     attorneyFees
   ]);
+  const handleVeteranIsNotClaimant = (value) => {
```
Comment (nanotone) [inline client/app/intake/components/SelectClaimant.jsx:86]: I added this tiny intermediate function to translate between the on/off veteran-not-claimant UI, and the claimantType field in the Redux store. I wanted to reduce the booleanness as much as possible, down to just this component in charge of the radio buttons.

File: client/app/intake/constants.js [hunk 8]
```diff
@@ -170,6 +170,36 @@ export const REVIEW_OPTIONS = {
   }
 };
 
+export const REVIEW_DATA_FIELDS = {
```
Comment (nanotone) [inline client/app/intake/constants.js:173]: The new `validateReviewData` and the existing `prepareReviewData` ended up being pretty similar, both revolving around a mapping between snakes and camels. Also, both having nearly zero real logic. So on @jcq's suggestion, I extracted all that stuff into a constant here.

Notice how this starts to look like a controller schema 😈

File: client/app/intake/util/index.js [hunk 9]
```diff
@@ -1,10 +1,10 @@
 import _ from 'lodash';
-import { REVIEW_OPTIONS, CLAIMANT_ERRORS } from '../constants';
+import { REVIEW_OPTIONS, REVIEW_DATA_FIELDS, CLAIMANT_ERRORS } from '../constants';
 import DATES from '../../../constants/DATES';
 import { formatDateStr } from '../../util/DateUtil';
 
 export const getBlankOptionError = (responseErrorCodes, field) => (
-  (_.get(responseErrorCodes[field], 0) === 'blank') && 'Please select an option.'
+  _.get(responseErrorCodes[field], 0) === 'blank' ? 'Please select an option.' : null
```
Comment (nanotone) [inline client/app/intake/util/index.js:7]: Just quelling a React warning, because the `&&` operator was spitting out string-or-boolean which proptypes was unhappy about. It's fine with string-or-null though 🤷

File: client/app/intake/util/index.js [hunk 10]
```diff
@@ -89,45 +89,32 @@ export const formatSearchableDropdownOptions = (options) => {
...
+    }
+  }
+  if (intakeData.receiptDate && intakeData.receiptDate > (new Date).toISOString()) {
+    errorCodes.receipt_date = ['in_future'];
+  }
+  if (['dependent', 'attorney'].includes(intakeData.claimantType) && !intakeData.claimant) {
+    errorCodes.claimant = ['blank'];
+  }
+  return (Object.keys(errorCodes).length ? errorCodes : null);
+};
```
Comment (nanotone) [inline client/app/intake/util/index.js:111]: Here's the frontend validation function! I did squeeze a couple custom checks in here, because it'd be nice for future receipt dates and missing claimants to get instantaneous error messages too.

The validation function returns an errorCodes object that's formatted exactly like the backend would return. That way, frontend and backend validation can use the same reducer to interpret errors codes as error messages.

The feature tests also expect to see many error messages simultaneously. This helped me get away with making fewer changes to the feature tests.

File: spec/support/intake_helpers.rb [hunk 11]
```diff
@@ -16,14 +16,24 @@ def select_form(form_name)
...
   def start_higher_level_review(
     test_veteran,
     receipt_date: 1.day.ago,
     claim_participant_id: nil,
     legacy_opt_in_approved: false,
-    veteran_is_not_claimant: false,
     benefit_type: "compensation",
-    informal_conference: false
+    informal_conference: false,
+    no_claimant: false
```
Comment (nanotone) [inline spec/support/intake_helpers.rb:36]: The first of a slew of setup methods, that could all use some tech debt cleanup.

With this one, the method originally took both a `claim_participant_id` (nil if veteran, present if dependent) and a `veteran_is_not_claimant`. Some tests relied on the claimant object not being created yet, other tests don't care. Some tests were okay with those two arguments disagreeing... until claimant subclasses entered the picture.

This was step 1 of a long journey of tidying up 😓

File: app/models/decision_review_intake.rb [hunk 12]
```diff
@@ -104,4 +125,34 @@ def after_validated_pre_start!
...
+    fail Caseflow::Error::MustImplementInSubclass
+  end
+  # :nocov:
+
+  def review_params
+    params = request_params.permit(*review_param_keys)
+    params[:veteran_is_not_claimant] = veteran_is_not_claimant
```
Comment (jcq) [inline app/models/decision_review_intake.rb:155]: I assume this use of `veteran_is_not_claimant` remains in order to somewhat minimize the additional code changes?

How about its use in `claimant_required_error` ? Is that still relevant?

File: app/models/higher_level_review_intake.rb [hunk 13]
```diff
@@ -16,14 +16,7 @@ def ui_hash
...
-  def review_params
-    request_params.permit(
-      :receipt_date,
-      :informal_conference,
-      :same_office,
-      :benefit_type,
-      :veteran_is_not_claimant,
-      :legacy_opt_in_approved
-    )
+  def review_param_keys
```
Comment (jcq) [inline app/models/higher_level_review_intake.rb:19]: I like these refactors

File: app/schemas/intakes_schemas.rb (hunk 3 above)
Comment (jcq) [reply app/schemas/intakes_schemas.rb:16]: And remind me (and/or for the record), what happens if this endpoint received `veteran_is_not_claimant` in a request? Does this schema do anything to prevent/flag that, or is that handling only at the level of the permitted params?

File: app/serializers/intake/decision_review_serializer.rb [hunk 14]
```diff
@@ -5,6 +5,10 @@ class Intake::DecisionReviewSerializer
   set_key_transform :camel_lower
 
   attribute :claimant, &:claimant_participant_id
```
Comment (jcq) [inline app/serializers/intake/decision_review_serializer.rb:7]: I definitely don't want any more scope creep, but after all the changes we've made, I kind of wish we could refactor everything to return a `claimant` object that contains the `participant_id`, `type`, etc — it's a bit weird to still refer to the participant ID simply as `claimant`

Ah well!

File: client/app/intake/actions/decisionReview.js [hunk 15]
```diff
@@ -12,6 +12,16 @@ export const submitReview = (intakeId, intakeData, intakeType) => (dispatch) =>
     meta: { analytics }
   });
 
+  const validationErrors = validateReviewData(intakeData, intakeType);
+  if (validationErrors) {
```
Comment (jcq) [inline client/app/intake/actions/decisionReview.js:16]: How did you avoid getting dinged for `padding-line-between-statements` eslint rule here? 🤔

File: client/app/intake/constants.js (hunk 8 above)
Comment (jcq) [reply client/app/intake/constants.js:173]: I think I understand how you're using this for a dual purpose, but might be missing something. Could this not have been accomplished arrays of the required fields keyed to the intakeType and usage of util fns to convert to/from snake case as necessary?

File: app/models/decision_review_intake.rb (hunk 12 above)
Comment (nanotone) [reply app/models/decision_review_intake.rb:155]: Yeah the decoupling is not 100% there yet. Consider `request_params` to be HTTP request params, and `review_params` (the return value) to be attributes bound for the DecisionReview record. I haven't gotten rid of `veteran_is_not_claimant` on that model, and it may be a while before we can.

The one in `claimant_required_error` is coupled to ClaimantValidator. That one seems somewhat easier to get rid of, and yeah we do still need it, to report missing participant IDs (you mention the naming mismatch later on).

File: spec/support/intake_helpers.rb (hunk 11 above)
Comment (leikkisa) [reply spec/support/intake_helpers.rb:36]: :nod:

File: app/schemas/intakes_schemas.rb [hunk 16]
```diff
@@ -15,18 +15,15 @@ def create
...
-               included_in?: DOCKET_OPTIONS,
-               doc: "Appeals only"
+        string :benefit_type, optional: true, doc: "not applicable to Appeals"
+        string :docket_type, optional: true, included_in?: DOCKET_OPTIONS, doc: "Appeals only"
         string :claimant, optional: true, nullable: true
+        string :claimant_type, optional: true, included_in?: %w[veteran dependent attorney other]
+        string :claimant_notes, optional: true, nullable: true, doc: "Appeals only"
         string :payee_code, optional: true, nullable: true
-        bool :informal_conference, optional: true, nullable: true, doc: "HLRs only"
+        bool :informal_conference, optional: true, doc: "HLRs only"
```
[the remark this answers is not in the document]
Comment (nanotone) [reply app/schemas/intakes_schemas.rb:24]: Yep, as long as the client omits `informal_conference` for the other review types -- which it does clearly as per that big constant object.

Removing `nullable: true` wherever appropriate allows the API call to convey its intent more clearly. Either `informal_conference` is applicable (HLR) and must be true or false, or it's not applicable and should be left out entirely. A null value (the client doesn't know / the user didn't select anything) shouldn't be interchangeable with omission.

File: app/serializers/intake/decision_review_serializer.rb (hunk 14 above)
Comment (nanotone) [reply app/serializers/intake/decision_review_serializer.rb:7]: Oh, that sounds like a great idea.

In terms of moving parts, I think of it as a renaming... where new names happen to include periods denoting JSON nesting. But it would still be a relatively self-contained change, so let's do that in a separate PR.

File: client/app/intake/util/issues.js [hunk 17]
```diff
@@ -3,20 +3,26 @@ import { formatDateStr } from '../../util/DateUtil';
 import DATES from '../../../constants/DATES';
 import { FORM_TYPES } from '../constants';
 
-const getNonVeteranClaimant = (intakeData) => {
-  const claimant = intakeData.relationships.filter((relationship) => {
-    return relationship.value === intakeData.claimant;
-  });
+const getDependentClaimant = (intakeData) => {
```
[the remark this answers is not in the document]
Comment (nanotone) [reply client/app/intake/util/issues.js:6]: Yeah, I wonder if the distinction between `util/index` and `util/issues` was originally based on whether it was used for the review page or the add-issues page. These functions are claimant-centric, but used only on the add-issues page to display the summary of the submitted form.

File: app/models/decision_review_intake.rb [hunk 18]
```diff
@@ -52,6 +52,27 @@ def build_issues(request_issues_data)
...
+      claimant.notes = request_params[:claimant_notes]
+      claimant.save!
+    end
+    update_person!
+  end
+
+  # :nocov:
+  def need_payee_code?
+    fail Caseflow::Error::MustImplementInSubclass
+  end
```
[the remark this answers is not in the document]
Comment (nanotone) [reply app/models/decision_review_intake.rb:73]: By similar addition, do you mean adding a `MustImplementInSubclass` abstract method in the superclass, for the `review!` method?

It was already there, in the base Intake class 😎 (not in DRIntake, because RAMP has it too)

File: app/models/intake.rb [hunk 19]
```diff
@@ -202,7 +202,8 @@ def ui_hash
       veteran_is_deceased: veteran&.deceased?,
       completed_at: completed_at,
       relationships: veteran&.relationships&.map(&:serialize),
-      processed_in_caseflow: detail.try(:processed_in_caseflow?)
+      processed_in_caseflow: detail.try(:processed_in_caseflow?),
+      claimantType: detail.try(:claimant_type)&.sub(/Claimant$/, "")&.downcase
```
[the remark this answers is not in the document]
Comment (nanotone) [reply app/models/intake.rb:206]: If you look at the happy-path for the review endpoint in `IntakesController`, it gets sent back to the frontend after a successful review submission. So this `claimantType` gets used by the code in `util/issues` to display claimant-related stuff on the add-issues page.

File: app/models/decision_review.rb [hunk 20]
```diff
@@ -156,6 +156,10 @@ def claimant_participant_id
     claimant&.participant_id
   end
 
+  def claimant_type
```
[the remark this answers is not in the document]
Comment (nanotone) [reply app/models/decision_review.rb:159]: Oh yes, that makes a lot of sense to me. I think the red herring here was the similarity between `claimant_type` and `claimant&.type`, but that's just a leak of STI implementation details (that it uses the `type` column).

I went ahead and did the rename; it wasn't bad at all. Now `claimant_type` always refers to the frontend version, `claimant_class` refers to the class (mostly used in tests) and `claimant_class_name` refers to what claimant_type used to. And as a bonus, the `downcase` code got DRYed up!

File: client/app/intake/constants.js (hunk 8 above)
Comment (nanotone) [reply client/app/intake/constants.js:173]: @jcq You're absolutely right... or maybe 99% right? Not using a camel/snake converter definitely made it more verbose, and I'm not sure why that slipped my mind.

What would be an effective JSON structure for capturing both required and non-required field names, though? It's entirely possible that there's an obvious one, and I'm just not thinking of it due to having stared at this for too long.

File: app/schemas/intakes_schemas.rb (hunk 3 above)
Comment (nanotone) [reply app/schemas/intakes_schemas.rb:16]: In its current state, the controller schema drops `veteran_is_not_claimant` from the params (and reports it to the Rails log) before passing it along to the controller.

**Comment by nanotone** [comment]
@leikkisa @jcq Thanks for the 👀 👀! I know it was a lot.

The one remaining issue I want to sort out is the `veteran_is_not_claimant` thing. First, to make sure we're on the same page, this is a summary of what's going on:
- The `veteran_is_not_claimant` DB column (on the 3 separate review tables) is not going away.
- For each review, this column will have value false if claimant type is `veteran`, true otherwise. The PR currently enforces this logic in the frontend, when interpreting the user's actions upon the component.
- The PR also proposes nixing the `veteran_is_not_claimant` field from the API. The controller schema would henceforth drop it entirely. So to figure out what gets saved to the DB, the backend would infer the value from the `claimant_type` param.

Open questions that we might just disagree on philosophically, and that's okay:
- Should the API have two fields when one can be derived from the other? (Yang prefers no)
- Should the frontend be allowed to submit an obviously inconsistent request, so the backend can detect it as a sign of a bug? (Yang prefers no, it sounds like Sally prefers yes?)
- Should the DB try to store the state of the UI, as experienced by the user? (Sally prefers yes)

Hopefully that's an accurate characterization, and if so, I think a good way to go might be to allow the API field back in, but do input validation and error returning as early as possible. But note that
- The frontend can't be expected to handle this error (because we can't predict the cause)
- A feature test can't cover it. Ideally a controller test would, but a model test could, too.

And if this discussion is about potential bugs rather than existing bugs in the current PR, I'm also inclined to say it should be deferred to a future PR.

**Comment by jcq** [comment]
@nanotone @leikkisa 
Just so I understand the considerations... what would be a scenario where capturing a boolean for `veteran_is_not_claimant` would enlighten (or disagree) with the claimant selection that the user has made? 

In general I support capturing/recording as much info about a user's actions/intentions as possible, but I also think it's possible to have UI elements that simply manipulate the UI in order to open up other user interactions that we do capture.

**Comment by nanotone** [comment]
My take is that there is no known scenario where `veteran_is_not_claimant` would disagree with the claimant selection. Rather, by capturing this boolean from the point that the user interacts with it (those first two radio buttons) and passing it along, we might detect bugs where the two end up being inconsistent with one another.

Ideally, `veteranIsNotClaimant` would always just be dependent on `claimantType != 'veteran'`, and I suspect that solidifying this dependency could alleviate the removal of the API param. But I wasn't sure how to engineer that on the frontend.

**Comment by jcq** [comment]
> Ideally, `veteranIsNotClaimant` would always just be dependent on `claimantType != 'veteran'`, and I suspect that solidifying this dependency could alleviate the removal of the API param. But I wasn't sure how to engineer that on the frontend.

If needed only within a given component, this is a good use case for the `useMemo` hook — it will memoize the value and only recompute if the specified input(s) change (in this case, `caseType`).

At the Redux level, this would be a good use case for a [memoized selector](https://redux.js.org/recipes/computing-derived-data).  (This would mainly be for brevity, as there's not much of a performance hit on a simple equality check.)
